### ISSUE TITLE
TransfiniteInterpolationManifold should copy (coarse) triangulation

### ISSUE THREAD

**Issue by tamiko** [dealii/dealii#11055]
I am currently having troubles with a `TransfiniteInterpolationManifold` and the following use case (simplified):

```
constexpr unsigned int dim = 2;

Triangulation<dim> coarse_triangulation;
GridGenerator::hyper_cube(coarse_triangulation);

coarse_triangulation.set_all_manifold_ids(9);
dealii::TransfiniteInterpolationManifold<2> transfinite;
transfinite.initialize(coarse_triangulation);
coarse_triangulation.set_manifold(9, transfinite);

Triangulation<dim> triangulation;
triangulation.copy_triangulation(coarse_triangulation);

triangulation.refine_global(1);
```

This (seemingly innocuous code) fails for two reasons:

(a) It will trigger a (somewhat performance critical) check in `TransfiniteInterpolationManifold<dim, spacedim>::get_possible_cells_around_points()`:
```
for (; cell != endc; ++cell)
  // only consider cells where the current manifold is attached
  if (&cell->get_manifold() != this)
    continue;
```
Here, the problem is that the `copy_triangulation` operation creates a bona fide copy of the manifold object and thus `&cell->get_manifold()` and `this` won't be the same object any more.

(b) The manifold stores a reference to the triangulation, so in above code suppose that I would retain the `triangulation` object but `coarse_triangulation` gets destroyed, the manifold can no longer function because it loses access to the coarse triangulation it was originally initialized on.

The underlying core problem in my opinion is that the `TransfiniteInterpolationManifold` actually stores a reference to the triangulation (that it must be attached to). I undestand why we are doing this - it is the a straight forward solution for the problem that the `TransfiniteInterpolationManifold` must be able to access the coarse grid and all attached manifolds.

Does something speak against copying the coarse grid information, i.e., all relevant coarse cells of the triangulation in addition to all relevant manifolds when `initialize` is called? This would solve both problems above in a (for the user) very convenient manner.

(((
Just in case someone is curious why I want to have use case (a): I need to create a coarse grid with some anisotropic pre refinement, a call to`flatten_triangulation`, and finally copying the final result into a (p4est) distributed triangulation. I have a heavily curved mesh and I want to have a transfinite interpolation between the outer curved boundary and an inner curved boundary (that is an obstacle). Recreating a TransfiniteInterpolationManifold for the (final) coarse mesh leads to very suboptimal results because coarse cells don't reach both boundaries any more.
)))

**Comment by tamiko** [comment]
@kronbichler, @luca-heltai *ping* :-)

**Comment by kronbichler** [comment]
I think having a copy of the triangulation would also solve the problem that #10971 tries to solve, so it would be two birds with one stone.

**Comment by tamiko** [comment]
I will have a look at it this evening :-)

**Comment by tamiko** [comment]
Another pragmatic solution for the problem in (a) could be to not `clone()` the manifold but to actually return a shared_ptr to one instance of a manifold object. That could be realized by adding a `set_manifold()` to the triangulation that takes a shared_ptr to a manifold as second argument and to store shared pointers to manifolds inside the triangulation.

With that little change, we should be able to retain full compatibility, i.e., manifolds can be supplied as objects (and get copied). But the user (and we for some cases in the library) can also decide to deliberately *not* copy a manifold but just to pass the shared pointer.

That would resolve (a), which is my main concern.

**Comment by tamiko** [comment]
@luca-heltai @kronbichler @jppelteret *ping* What do you think of this approach?

**Comment by kronbichler** [comment]
I would not mind having `shared_ptr` objects for the manifolds rather than 

What we are discussing here, the requirement stated in #10971 and the case of big coarse meshes in #11058 makes me think that we cannot avoid having multiple entry points to the transfinite interpolation:
- The default setup we have right now, setting a pointer to an externally organized triangulation
- The case of `TransfiniteInterpolation` holding its own `Triangulation` object. I would love if we only need the case of a single coarse cell. This would have the advantage of avoiding ambiguity in the identification of the coarse cell to operate on and make the whole algorithmic part more robust. The downside is that we would need to handle many different `TransfiniteInterpolation` objects, so we would need a library function `triangulation.set_transfinite_interpolation_where_necessary()` or something like that. The question is whether we can keep this stable across dynamic mesh adaptations and other manipulations.

I think the second function would solve the problem discussed in #10971 and also the issue discussed in the top post here; one would simply add a transfinite interpolation object to each panel and attach that object to a second triangulation. What do other think?

**Comment by tamiko** [comment]
I like the idea to restrict the transfinite interpolation to just a single coarse cell. This has the potential to simplify the current logic quite a bit. But there are still a couple of questions in that regard. The biggest one I am having is whether we then also copy the cell information (i.e., coordinates and all relevant manifolds)?

Also, I kind of start to like the `triangulation.set_transfinite_interpolation_where_necessary()` idea. After all, the idea behind `TransfiniteInterpolationManifold` is a bit of a special one: We do not have any geometry information in form of a manifold that is independent of the triangulation. Instead, all we really want to express is "hey, you now curvature on the left and on the right, please do something intuitively meaningful in between".

Shall we try to have another short developer zoom to discuss that in more detail? I would like to understand the problem a bit more and reach a consensus before hacking code together :-D

**Comment by kronbichler** [comment]
I think having a short meeting with all people interested would be great to see other opinions on this topic and see how we can set things up and get all requirements go together.

**Comment by kronbichler** [comment]
> This has the potential to simplify the current logic quite a bit. But there are still a couple of questions in that regard. The biggest one I am having is whether we then also copy the cell information (i.e., coordinates and all relevant manifolds)?

Yes, we would copy the cell information in that case to meet the requirements expressed above, so the important part is to keep this consistent with the actual mesh.

**Comment by jppelteret** [comment]
> I think having a short meeting with all people interested would be great to see other opinions on this topic and see how we can set things up and get all requirements go together.

Ditto!

> Yes, we would copy the cell information in that case to meet the requirements expressed above, so the important part is to keep this consistent with the actual mesh.

Another thing that we must (perhaps) be mindful of is what happens if the mesh gets updated after the TFI manifold is attached, e.g. if mesh smoothing is applied. There might be other legitimate operations that involve modifying the original triangulation, and I'd be curious to know whether this if this is an issue that we needs consideration and what could be done to keep the Triangulation and any of its copies (whole, or in part) in sync.

**Comment by peterrum** [comment]
This is an important topic. Let's postpone it for now. It is already on the agenda of the next workshop.

**Comment by peterrum** [comment]
@tamiko I guess this issue will be relevant again, once you move your airfold geometry generator to deal.II?

**Comment by bangerth** [comment]
I don't think this is going to make it for 9.4. Re-tagging for 9.5.

**Comment by tamiko** [comment]
@kronbichler This year we will fix it :-)

**Comment by tamiko** [comment]
@kronbichler https://github.com/conservation-laws/ryujin/pull/373
